packages/driver/src/cy/net-stubbing/events/response.ts

```diff
@@ -100,12 +100,12 @@ export const onResponse: HandlerFn<CyHttpMessages.IncomingResponse> = async (Cyp
     responseSent = true
 
     // copy changeable attributes of userRes to res
-    _.merge(res, _.pick(userRes, SERIALIZABLE_RES_PROPS))
+    mergeWithArrayBuffer(res, _.pick(userRes, SERIALIZABLE_RES_PROPS))
```

Comment:
@flotwig this is a nasty one. The explanation I can come up with is this: The socket.io parser [uses](https://github.com/socketio/socket.io-parser/blob/af1b23ca85fb64a0d7a050abd0362c5f632ce429/lib/is-binary.ts#L27) `instanceof ArrayBuffer` to check whether a part of the object to be encoded is an ArrayBuffer. Now the place where the ArrayBuffer coming from the real server response and the place where a stubbed ArrayBuffer response are constructed are not in the same scope and thus the prototype of the respective ArrayBuffers are not the same, i.e. `instanceof` returns `false` even though the object provided is an ArrayBuffer but from a different context (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof#instanceof_and_multiple_context_e.g._frames_or_windows). So if you just use merge here you get a very puzzling behaviour:

- if you intercept a binary response but do not touch the body, everything works as expected (ArrayBuffer is detected as binary from socket.io parser and encoded correspondingly)
- if you intercept a binary response and replace the body by another ArrayBuffer you construct inside of your e2e test, it is encoded as text and the Cypress server receives `{}` instead of the encoded binary content.

The possible solutions I see are:

- Create a new ArrayBuffer and deep copy the binary values (implemented in [utils.ts](https://github.com/cypress-io/cypress/pull/16755/files#diff-b0318835d7e6e6b3346d8b261b8647d0848c58591e3242a87af81483cd52380b)). I'm not 100% sure though if the copying via `Uint8Array` works for all scenarios.
- Overwrite the prototype with [Object.setPrototypeOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf) (also implemented in [utils.ts](https://github.com/cypress-io/cypress/pull/16755/files#diff-b0318835d7e6e6b3346d8b261b8647d0848c58591e3242a87af81483cd52380b))
- Write a [custom parser](https://socket.io/docs/v3/custom-parser/) for socket.io that for example uses `_.isArrayBuffer` instead of `instanceof`.

Please let me know what you prefer to do here (I tend to go for the second option) so that I can adapt my PR accordingly.

Reply:
@simfeld Cypress already patches `socket.io-parser` to use `CircularJSON` to allow us to pass circular objects through `socket.io`: https://github.com/cypress-io/cypress/blob/77880f1a0f0cecaee92e1f271cede452cc133287/packages/socket/patches/socket.io-parser+4.0.4.patch#L1-L162

It would be good to extend the patch so that we can have consistent `socket.io` behavior throughout the codebase. There are tests for the patched parser in `packages/socket/test/socket_spec.js`. The patch is applied with the `patch-package` file, so to update it, make changes in `packages/socket/node_modules/socket.io-parser`, and run `yarn patch-package socket.io-parser` to update the diff.

Go with whatever makes the implementation the easiest to understand - from my understanding of your comment, this would probably be updating the patch, since it would remove the need for `mergeWithArrayBuffer`?

Reply:
@flotwig thanks for the hint, that's clearly a cleaner solution. See 67e00da. Note that both my repro tests (https://github.com/simfeld/cypress-intercept-binary) and the tests I added in this PR also pass with the `engio.io-parser` patch _only_. I'm not quite sure in what scenarios `socket.io-parser` is used, but I included the same fix there anyway.